### DISCUSSION TITLE
Need help: MATERIALIZED VIEW always returns 1

### DISCUSSION THREAD

**Discussion by nikrovir** [ClickHouse/ClickHouse#63490] [Q&A]
Hi all.
I just recently started studying ClickHouse, so I ask you to be lenient with the question.

Created a table to store all article visitors and record viewing progress:

```sql
CREATE TABLE IF NOT EXISTS publication_visitors
(
    progress_id UUID,
    publication_id UUID,
    visitor_id UUID,
    action_type VARCHAR,
    seconds UInt64,
    percentage Float32,
    created_at DateTime('UTC') DEFAULT now()
) ENGINE = MergeTree() ORDER BY (created_at);
```

I also created a MATERIALIZED VIEW table for it:

```sql
CREATE MATERIALIZED VIEW IF NOT EXISTS publication_visitors_total
ENGINE = AggregatingMergeTree() ORDER BY (publication_id, created_at_hour) POPULATE
AS SELECT
    publication_id,
    toStartOfHour(created_at) as created_at_hour,
    count(DISTINCT visitor_id) as views_count
FROM publication_visitors
WHERE action_type = 'CREATE'
GROUP BY
    publication_id,
    created_at_hour;
```

I can’t understand why in the MATERIALIZED VIEW table `views_count`, even after recalculation, is always equal to 1 for each unique `publication_id` and `created_at_hour`? To make it clearer, let me clarify: for example, I add 2 records with obviously unique data. If I run a query from my MATERIALIZED VIEW:

```sql
SELECT
    publication_id,
    toStartOfHour(created_at) as created_at_hour,
    count(DISTINCT visitor_id) as views_count
FROM publication_visitors
WHERE action_type = 'CREATE'
GROUP BY
    publication_id,
    created_at_hour;
```

then I get `views_count` with the value **2**. However, the MATERIALIZED VIEW table stores the value **1** even after recalculation. Why does this work this way and how can it be fixed?

**Comment by UnamedRus** [comment]
>     count(DISTINCT visitor_id) as views_count

Wrong for MV, in ClickHouse MV it's more insert trigger.
Check article:
https://kb.altinity.com/altinity-kb-schema-design/materialized-views/

There is refresh MV(experimental), which act like "normal" MV from PosgreSQL for ex

https://clickhouse.com/docs/en/sql-reference/statements/create/view#refreshable-materialized-view
